include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -0,0 +1,923 @@
+  SizeType num_dense;
+  SizeType num_lowhalf;
+  SizeType num_deflated;
+};
+
+// Auxialiary matrix and vectors used for the D&C algorithm
+template <class T>
+struct WorkSpace {
+  WorkSpace(WorkSpace&&) = default;
+  WorkSpace& operator=(WorkSpace&&) = default;
+
+  // Extra workspace for Q1', Q2' and U1' if n2 > n1 or U2' if n1 >= n2 which are packed as follows:
+  //
+  //   ┌──────┬──────┬───┐
+  //   │  Q1' │  Q2' │   │
+  //   │      │      │   │
+  //   ├──────┤      │   │
+  //   │      └──────┘   │
+  //   │                 │
+  //   ├────────────┐    │
+  //   │     U1'    │    │
+  //   │            │    │
+  //   └────────────┴────┘
+  //
+  Matrix<T, Device::CPU> mat1;
+  Matrix<T, Device::CPU> mat2;
+
+  // Holds the values of the deflated diagonal sorted in ascending order
+  Matrix<T, Device::CPU> dtmp;
+  // Holds the values of Cuppen's rank-1 vector
+  Matrix<T, Device::CPU> z;
+  // Holds the values of the rank-1 update vector sorted corresponding to `d_defl`
+  Matrix<T, Device::CPU> ztmp;
+
+  // Steps:
+  //
+  // 1. Sort index based on diagonal values in ascending order. This creates a map sorted to inital indices
+  //
+  //        initial <--- pre_sorted
+  //
+  // 2. Sort index based on column types such that all deflated entries are at the end
+  //
+  //        pre_sorted <--- deflated
+  //
+  // 3. Sort index based on updated diagonal values in ascending order. The diagonal conatins eigenvalues
+  //    of the deflated problem and deflated entreis from the initial diagonal
+  //
+  //        deflated <--- post_sorted
+  //
+  // 4. Sort index based on column types such that matrices `Q` and `U` are in matrix multiplication form.
+  //
+  //        post_sorted <--- matmul
+  //
+  Matrix<SizeType, Device::CPU> i1;
+  Matrix<SizeType, Device::CPU> i2;
+  Matrix<SizeType, Device::CPU> i3;
+
+  // Assigns a type to each column of Q which is used to calculate the permutation indices for Q and U
+  // that bring them in matrix multiplication form.
+  Matrix<ColType, Device::CPU> c;
+  Matrix<ColType, Device::CPU> ctmp;
+};
+
+template <class T>
+WorkSpace<T> initWorkSpace(const matrix::Distribution& ev_distr) {
+  LocalElementSize vec_size(ev_distr.size().rows(), 1);
+  TileElementSize vec_tile_size(ev_distr.blockSize().rows(), 1);
+  WorkSpace<T> ws{Matrix<T, Device::CPU>(ev_distr),
+                  Matrix<T, Device::CPU>(ev_distr),
+                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<ColType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<ColType, Device::CPU>(vec_size, vec_tile_size)};
```

Comment:
Designated initializers would be nice to have here :disappointed: Without designated initializers I sometimes find
```
mystruct s;
s.a = some_a;
s.b = some_b;
...
```
nicer because it makes it (more) obvious if you've put a variable in the wrong member variable, but I don't mind the way it is right now either.

Reply:
I agree this is not ideal but I suggest we leave it for a future refactoring.